petric.py

```diff
@@ -88,26 +88,51 @@ def __call__(self, algo: Algorithm):
         log.debug("...logged")
 
 
+class QualityMetrics(ImageQualityCallback):
+    """From https://github.com/SyneRBI/PETRIC/wiki#metrics-and-thresholds"""
+    def __init__(self, reference_image, whole_object_mask, background_mask, **kwargs):
+        super().__init__(reference_image, **kwargs)
+        self.whole_object_indices = np.where(whole_object_mask == 1)
+        self.background_indices = np.where(background_mask == 1)
+        self.ref_im_arr = reference_image.as_array()
+
+    def __call__(self, algorithm):
+        iteration = algorithm.iteration
+        if iteration % algorithm.update_objective_interval != 0 and iteration != algorithm.max_iteration:
+            return
+
+        for filter_name, filter_func in self.filter.items():
+            if filter_func is None:
+                filter_func = lambda x: x
+            test_im, ref_im = (filter_func(img_data).as_array() for img_data in (algorithm.x, self.reference_image))
```

Comment:
is this what you mean Kris? i.e.

```py
def filter_fun(x: ImageData) -> ImageData
```

or do you want

```py
def filter_fun(x: ImageData) -> ndarray
```

```suggestion
            if filter_func is None:
                filter_func = lambda x: x.as_array()
            test_im, ref_im = (filter_func(img_data) for img_data in (algorithm.x, self.reference_image))
```

Reply:
I think the current option is fine. thx

Reply:
You should always use `algorithm.get_ouput()` unless you are sure that `.x` represents the current iterate.